python/ray/_private/services.py

```diff
@@ -1344,24 +1340,32 @@ def start_api_server(
         )
 
         # Retrieve the dashboard url
-        gcs_client = GcsClient(address=gcs_address, cluster_id=cluster_id_hex)
-        ray.experimental.internal_kv._initialize_internal_kv(gcs_client)
         dashboard_url = None
         dashboard_returncode = None
-        for _ in range(200):
-            dashboard_url = ray.experimental.internal_kv._internal_kv_get(
-                ray_constants.DASHBOARD_ADDRESS,
+        dashboard_start_tiemout_time = (
+            time.perf_counter() + ray_constants.DASHBOARD_START_TIMEOUT_S
+        )
+        while True:
```

Comment:
@edoakes any ideas of how to test this? 

I can either make the dashboard slow to start and test the entire `start_api_server` function or only test the `while True` block by making it a function.

Reply:
in an ideal world:

- inject a fake clock
- inject a fake `gcs_client` and process management interface
- write the tests directly against `start_api_server`

Reply:
the currently-pragmatic solution would be to pull the logic into a utility and pass in a callback that checks if dashboard is up, then test that w/ a different callback

Reply:
How about this?
- As Ed suggested, extract the wait loop into a helper so you can inject a fake GcsClient/KV stub.
- Simulate slow startup by having your stub return None for a few iterations, then a URL - verify the timeout path and normal path.